README.md
jaod
====



[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![cran checks](https://cranchecks.info/badges/worst/jaod)](https://cranchecks.info/pkgs/jaod)
[![R-check](https://github.com/ropensci/jaod/workflows/R-check/badge.svg)](https://github.com/ropensci/jaod/actions?query=workflow%3AR-check)
[![codecov.io](https://codecov.io/github/ropensci/jaod/coverage.svg?branch=master)](https://codecov.io/github/ropensci/jaod?branch=master)
[![rstudio mirror downloads](https://cranlogs.r-pkg.org/badges/jaod)](https://github.com/r-hub/cranlogs.app)
[![cran version](https://www.r-pkg.org/badges/version/jaod)](https://cran.r-project.org/package=jaod)

`jaod` - R client for [Directory of Open Access Journals API](https://doaj.org/api/v2/docs)

DOAJ API FAQ: <https://doaj.org/docs/faq/#apikey>

get started with the package docs: <https://docs.ropensci.org/jaod/>

## installation

stable CRAN version


```r
install.packages("jaod")
```

development version


```r
remotes::install_github("ropensci/jaod")
```


```r
library("jaod")
```

## Meta

* Please [report any issues or bugs](https://github.com/ropensci/jaod/issues).
* License: MIT
* Get citation information for `jaod` in R doing `citation(package = jaod')`
* Please note that this package is released with a [Contributor Code of Conduct](https://ropensci.org/code-of-conduct/). By contributing to this project, you agree to abide by its terms.


NEWS.md
jaod 0.3.0
==========

### MINOR IMPROVEMENTS

* fix to `jaod_article_search()` (#11)
* change base DOAJ API url to v2 (version 2) (#12)


jaod 0.2.0
==========

### MINOR IMPROVEMENTS

* package gains a vignette by Lluís Sancho (#2) (#7)
* now using `vcr` for tests (#8)
* added link to the FAQ for the DOAJ API in the readme and package level manual file (#9)


jaod 0.1.0
==========

### NEW FEATURES

* Released to CRAN.


tests/test-all.R
library("testthat")
library("jaod")

test_check("jaod")


tests/testthat/test-jaod_article.R
context("jaod_article")

test_that("jaod_article works", {
  vcr::use_cassette("jaod_article", {
    a <- jaod_article(id = "0005e11ec616453f854070069385e057")
  })

  expect_is(a, "list")
  expect_true("id" %in% names(a))
  expect_is(a$admin, "list")
  expect_is(a$last_updated, "character")
  expect_is(a$id, "character")
  expect_is(a$bibjson, "list")
  expect_is(a$bibjson$title, "character")
  expect_is(a$created_date, "character")
})

test_that("jaod_article many inputs", {
  vcr::use_cassette("jaod_article_many_inputs", {
    ids <- c("73f781ad57094b98a0b46d8903fdc2c2",
      "0005e11ec616453f854070069385e057")
    d <- lapply(ids, jaod_article)
  })

  expect_is(d, "list")
  expect_is(d[[1]], "list")
  expect_is(d[[2]], "list")
  expect_equal(length(d), 2)
})

test_that("jaod_article fails well", {
  skip_on_cran()

  # missing id
  expect_error(jaod_article(), "argument \"id\" is missing")

  # not found
  expect_error(jaod_article("asdf"), "HTTP 404", class = "error")

  # wrong type
  expect_error(jaod_article(5), "id must be of class character")
})


tests/testthat/test-jaod_journal.R
context("jaod_journal")

test_that("jaod_journal works", {
  vcr::use_cassette("jaod_journal", {
    a <- jaod_journal(id = "f3f2e7f23d444370ae5f5199f85bc100")
  })

  expect_is(a, "list")
  expect_true("id" %in% names(a))
  expect_is(a$admin, "list")
  expect_is(a$last_updated, "character")
  expect_is(a$id, "character")
  expect_is(a$bibjson, "list")
  expect_is(a$bibjson$title, "character")
  expect_is(a$created_date, "character")
})

test_that("jaod_journal many inputs", {
  vcr::use_cassette("jaod_journal_many_inputs", {
    ids <- c("f3f2e7f23d444370ae5f5199f85bc100",
             "9abfb36b06404e8a8566e1a44180bbdc")
    d <- lapply(ids, jaod_journal)
  })

  expect_is(d, "list")
  expect_is(d[[1]], "list")
  expect_is(d[[2]], "list")
  expect_equal(length(d), 2)
})

test_that("jaod_journal fails well", {
  skip_on_cran()

  # missing id
  expect_error(jaod_journal(), "argument \"id\" is missing")

  # not found
  expect_error(jaod_journal("asdf"), "HTTP 404", class = "error")

  # wrong type
  expect_error(jaod_journal(5), "id must be of class character")
})


R/jaod_journal_search.R
#' Search for journals
#'
#' @export
#' @inheritParams jaod_article_search
#' @return a list with metadata (`timestamp`, `page`, `pageSize`, `query`,
#' `total` (number of results), `next` and `last` (next page and last page),
#' and `results` (a tibble (data.frame)))
#' @section query:
#' You can search inside any field you see in the results or the schema.
#' See <https://doaj.org/api/v1/docs#specific_field_search> for more details.
#' For example, to search for all articles with abstracts containing the
#' word "shadow", you would do `bibjson.abstract:"shadow"`.
#'
#' Short-hand names are available for some fields. See
#' <https://doaj.org/api/v1/docs#short_field_names> for more details. For
#' example: `doi:10.3389/fpsyg.2013.00479`, `issn:1874-9496`, `license:CC-BY`,
#' and `title:hydrostatic pressure`
#' @examples \dontrun{
#' out <- jaod_journal_search(query = "bibjson.keywords:heritage")
#' out
#' out$results
#'
#' jaod_journal_search(query = "issn:1544-9173")
#' jaod_journal_search(query = "publisher:dove")
#'
#' # sorting
#' out <- jaod_journal_search(query = "bibjson.keywords:heritage",
#'   sort = "bibjson.eissn:desc")
#' out$results$bibjson.eissn
#' }
jaod_journal_search <- function(query, page = 1, pageSize = 10, sort = NULL,
                                ...) {

  assert(query, "character")
  assert(page, c('integer', 'numeric'))
  assert(pageSize, c('integer', 'numeric'))
  assert(sort, "character")
  res <- jaod_parse(
    jGET(
      file.path(doaj_base(), "search", "journals", query),
      query = jc(list(page = page, pageSize = pageSize, sort = sort)),
      ...
    )
  )
  res$results <- tibble::as_tibble(res$results)
  nmo <- c('timestamp', 'page', 'pageSize', 'query', 'total',
           'next', 'last', 'results')
  return(res[nmo])
}


R/jaod_article_search.R
#' Search for articles
#'
#' @export
#' @param query (character) query terms. See Details.
#' @param page (integer) Which page of the results you wish to see. Default: 1
#' @param pageSize (integer) How many results per page you wish to see.
#' Default: 10
#' @param sort (character) one of asc or desc, or sort by a specific field,
#' either asc or desc, see examples. If asc or desc not given, asc is used
#' @template curl
#' @return a list with metadata (`timestamp`, `page`, `pageSize`, `query`,
#' `total` (number of results), `next` and `last` (next page and last page),
#' and `results` (a tibble (data.frame)))
#' @section query:
#' You can search inside any field you see in the results or the schema.
#' See <https://doaj.org/api/v1/docs#specific_field_search> for more details.
#' For example, to search for all articles with abstracts containing the
#' word "shadow", you would do `bibjson.abstract:"shadow"`.
#'
#' Short-hand names are available for some fields. See
#' <https://doaj.org/api/v1/docs#short_field_names> for more details. For
#' example: `doi:10.3389/fpsyg.2013.00479`, `issn:1874-9496`, `license:CC-BY`,
#' and `title:hydrostatic pressure`
#' @examples \dontrun{
#' out <- jaod_article_search(query = "license:CC-BY")
#' out
#' out$results
#'
#' ## get DOIs or other article IDs
#' names(out$results$bibjson.identifier) <- out$results$id
#' do.call(rbind, out$results$bibjson.identifier)
#'
#' # more egs
#' jaod_article_search(query = "issn:1544-9173")
#' jaod_article_search(query = "publisher:dove")
#'
#' # sorting
#' res <- jaod_article_search(query="Pinus contorta", sort="year:desc",
#'   pageSize = 50)
#' res$results$bibjson.year
#' jaod_article_search(query="green", sort="bibjson.year:desc")
#' }
jaod_article_search <- function(query, page = 1, pageSize = 10, sort = NULL,
                                ...) {

  assert(query, "character")
  assert(page, c('integer', 'numeric'))
  assert(pageSize, c('integer', 'numeric'))
  assert(sort, "character")
  res <- jaod_parse(
    jGET(
      file.path(doaj_base(), "search", "articles", query),
      query = jc(list(page = page, pageSize = pageSize, sort = sort)),
      ...
    )
  )
  res$results <- tibble::as_tibble(res$results)
  nmo <- c('timestamp', 'page', 'pageSize', 'query', 'total',
           'next', 'last', 'results')
  nmo_exist <- nmo[nmo %in% names(res)]
  return(res[nmo_exist])
}
